### PR TITLE
Debug flags

### PR DESCRIPTION
**Pull request by guolinke**
for #2820 and #1588

- [x] add documents

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -171,10 +182,17 @@ if(MSVC)
...
+        SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /Od")
+    else()
+        SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /O2 /Ob2 /Oi /Ot /Oy")
+    endif()
 else()
-    SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fPIC")
-    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -funroll-loops")
+    if(NOT USE_DEBUG)
+      SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fPIC")
+      set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -funroll-loops")
```
Comment (StrikerRUS) [inline CMakeLists.txt]: Can we simplify it?
```suggestion
      SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fPIC -funroll-loops")
```

File: windows/LightGBM.vcxproj [hunk 2]
```diff
@@ -64,6 +75,11 @@
     <LibraryPath>$(VC_LibraryPath_x64);$(WindowsSDK_LibraryPath_x64);</LibraryPath>
     <TargetName>lightgbm</TargetName>
   </PropertyGroup>
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='DEBUG_DLL|x64'">
+    <IncludePath>..\include;$(VC_IncludePath);$(WindowsSDK_IncludePath);</IncludePath>
+    <LibraryPath>$(VC_LibraryPath_x64);$(WindowsSDK_LibraryPath_x64);</LibraryPath>
+    <TargetName>lightgbm</TargetName>
```
Comment (StrikerRUS) [inline windows/LightGBM.vcxproj]: To be consistent with `DLL|x64`?
```suggestion
    <TargetName>lib_lightgbm</TargetName>
```

File: windows/LightGBM.sln [hunk 3]
```diff
@@ -7,13 +7,16 @@ Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "LightGBM", "LightGBM.vcxpro
 EndProject
 Global
 	GlobalSection(SolutionConfigurationPlatforms) = preSolution
+	    Debug_DLL|x64 = Debug_DLL|x64
```
Comment (guolinke) [inline windows/LightGBM.sln]:
```suggestion
    Debug_DLL|x64 = Debug_DLL|x64
```

File: windows/LightGBM.sln [hunk 4]
```diff
@@ -7,13 +7,16 @@ Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "LightGBM", "LightGBM.vcxpro
 EndProject
 Global
 	GlobalSection(SolutionConfigurationPlatforms) = preSolution
+    Debug_DLL|x64 = Debug_DLL|x64
```
Comment (guolinke) [inline windows/LightGBM.sln]:
```suggestion
   Debug_DLL|x64 = Debug_DLL|x64
```

**Comment by guolinke** [comment]
thanks @jameslamb , will add related documentation when have time.

**Comment by guolinke** [comment]
@StrikerRUS Thank you so much! And it would be better!

**Comment by StrikerRUS** [comment]
@guolinke WDYT about compiling LightGBM in debug mode to check that everithing is OK at CI?
For example we can set debug here, because we cannot set it for `regular` job as we use artifacts from there in releases.
https://github.com/microsoft/LightGBM/blob/53137e2541691fa8bcae2917eb75b7ddf0ec98c8/.ci/test.sh#L141

**Comment by guolinke** [comment]
Good idea!

**Comment by StrikerRUS** [comment]
I added docs for new compilation options and set MPI job to be run in `Debug` mode in the latest commits.

**Pull request merged**